sdk/keyvault/keyvault-certificates/src/core/challengeBasedAuthenticationPolicy.ts

```diff
@@ -82,102 +108,126 @@ export class ChallengeBasedAuthenticationPolicy extends BaseRequestPolicy {
+    // or if the cached challenge has a different scope,
+    // we store the just received challenge and reset the cached token, to force a re-authentication.
+    if (!this.challengeCache.challenge?.equalTo(challenge)) {
+      this.challengeCache.setCachedChallenge(challenge);
+      this.tokenCache.setCachedToken(undefined);
+    }
 
-    // Ensure that we're about to use a secure connection
+    await this.loadToken(webResource);
+    return this._nextPolicy.sendRequest(webResource);
+  }
+
+  /**
+   * Applies the Bearer token to the request through the Authorization header.
+   * @param webResource Ongoing HTTP request.
+   */
+  public async sendRequest(webResource: WebResource): Promise<HttpOperationResponse> {
+    // Ensure that we're about to use a secure connection.
     if (!webResource.url.startsWith("https:")) {
       throw new Error("The resource address for authorization must use the 'https' protocol.");
     }
 
-    const originalBody = webResource.body;
+    // The next request will happen differently whether we have a challenge or not.
+    let response: HttpOperationResponse;
 
     if (this.challengeCache.challenge == undefined) {
-      // Use a blank to start the challenge
+      // If there's no challenge in cache, a blank body will start the challenge.
+      const originalBody = webResource.body;
       webResource.body = "";
+      response = await this._nextPolicy.sendRequest(webResource);
+      webResource.body = originalBody;
     } else {
-      // or use the cached token if we have one
-      await this.authenticateRequest(webResource);
+      // If we did have a challenge in memory,
+      // we attempt to load the token from the cache into the request before we try to send the request.
+      await this.loadToken(webResource);
+      response = await this._nextPolicy.sendRequest(webResource);
     }
 
-    const response = await this._nextPolicy.sendRequest(webResource);
-
-    if (response.status == 401) {
-      webResource.body = originalBody;
-
-      let www_authenticate = response.headers.get("WWW-Authenticate");
-
-      if (www_authenticate) {
-        // The challenge based authentication will contain both an authorization URI with a token,
-        // and the resource to which that token is valid against (also called the scope).
-        const { authorization, resource } = this.parseWWWAuthenticate(www_authenticate);
-        const challenge = new AuthenticationChallenge(authorization, resource + "/.default")
-
-        if (!challenge.equalTo(this.challengeCache.challenge)) {
-          this.challengeCache.setCachedChallenge(challenge);
-          this.tokenCache.setCachedToken(undefined);
-
-          await this.authenticateRequest(webResource);
-          return this._nextPolicy.sendRequest(webResource);
-        }
-        return response;
-      }
-      return response;
-    } else {
+    // If we don't receive a response with a 401 status code,
```

Comment:
can everything from here until the end be moved up after the line `webResource.body = originalBody` inside the body of the if? Since we didn't blank the body in the else, we know it will never respond with 401 and we can just return the response.

Reply:
Do we know it will never return 401? Seems like we are being extra cautious if we double check like this.

Reply:
Let's say we're already authenticated, but the server answers with 401 when our code tries to:

```ts
      await this.loadToken(webResource);
      response = await this._nextPolicy.sendRequest(webResource);
```

Seems possible! I'm guessing that if the challenge or token have some kind of expiration date, this might help us correctly re-authenticate.

Reply:
ah, sure

Reply:
(I'm seeing an A-zure pun and I don't know what to do with it 😄)